File: NetDeck/OctgnImport.m
//
//  OctgnImport.m
//  Net Deck
//
//  Created by Gereon Steffens on 16.04.14.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "OctgnImport.h"

@interface OctgnImport()

@property NSXMLParser* parser;
@property Deck* deck;
@property NSMutableString* notes;

@end

@implementation OctgnImport

-(Deck*) parseOctgnDeckFromData:(NSData *)data
{
    self.parser = [[NSXMLParser alloc] initWithData:data];
    self.parser.delegate = self;

    return [self parse];
}

-(Deck*) parseOctgnDeckWithParser:(NSXMLParser *)parser
{
    self.parser = parser;
    self.parser.delegate = self;
    
    return [self parse];
}

-(Deck*) parse
{
    self.deck = [[Deck alloc] init];
    
    if ([self.parser parse] && self.deck.role != NRRoleNone)
    {
        return self.deck;
    }
    else
    {
        return nil;
    }
}

-(void) parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName attributes:(NSDictionary *)attributeDict
{
    self.notes = nil;
    
    if ([elementName isEqualToString:@"card"])
    {
        NSString* qty = attributeDict[@"qty"];
        NSString* code = attributeDict[@"id"];
        
        if ([code hasPrefix:OCTGN_CODE_PREFIX] && code.length > 32)
        {
            Card* card = [CardManager cardByCode:[code substringFromIndex:31]];
            int copies = [qty intValue];
        
            if (card)
            {
                // NSLog(@"card: %d %@", copies, card.name);
                [self.deck addCard:card copies:copies];
                self.deck.role = card.role;
            }
        }
    }
    
    if ([elementName isEqualToString:@"notes"])
    {
        self.notes = [NSMutableString string];
    }
}

-(void) parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName
{
    if ([elementName isEqualToString:@"notes"])
    {
        self.deck.notes = self.notes;
    }
}

-(void) parser:(NSXMLParser *)parser foundCharacters:(NSString *)string
{
    [self.notes appendString:string];
}

@end

File: NetDeck/OctgnImport.swift
//
//  OctgnImport.swift
//  NetDeck
//
//  Created by Gereon Steffens on 19.02.16.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import Foundation
import SwiftyUserDefaults

final class OctgnImport: NSObject, XMLParserDelegate {
    
    private let deck = Deck(role: .none)
    private var notes: String?
    
    func parseOctgnDeckFromData(_ data: Data) -> Deck? {
        let parser = XMLParser(data: data)
        parser.delegate = self
        
        let ok = parser.parse()
        if ok && self.deck.role != .none {
            return self.deck
        } else {
            return nil
        }
    }
    
    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String: String]) {
        if elementName == "card" {
            if let qty = attributeDict["qty"], let id = attributeDict["id"] {
                if id.hasPrefix(Card.octgnPrefix) && id.count > 32 {
                    let index = id.index(id.startIndex, offsetBy: 31)
                    let cardCode = String(id[index...])
                    let code = Card.originalToRevised[cardCode] ?? cardCode
                    if let card = CardManager.cardBy(code), let copies = Int(qty) {
                        // NSLog(@"card: %d %@", copies, card.name);
                        self.deck.addCard(card, copies: copies)
                    }
                }
            }

        } else if elementName == "notes" {
            self.notes = ""
        }
    }
    
    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        if elementName == "notes" {
            self.deck.notes = self.notes
        }
    }
    
    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if self.notes != nil {
            self.notes! += string
        }
    }
}
